### PR TITLE
[libc] Refactor _build_gpu_objects cmake function.

### PR DESCRIPTION
**Pull request by lntue**

### CODE REVIEW

File: libc/cmake/modules/LLVMLibCObjectRules.cmake [hunk 1]
```diff
@@ -145,21 +145,78 @@ function(get_nvptx_compile_options output_var gpu_arch)
...
+  if(NOT ADD_GPU_OBJ_CXX_STANDARD)
+    set(ADD_GPU_OBJ_CXX_STANDARD ${CMAKE_CXX_STANDARD})
+  endif()
+
+  set(compile_options ${ADD_GPU_OBJ_COMPILE_OPTIONS})
+  # Derive the triple from the specified architecture.
+  if("${gpu_arch}" IN_LIST all_amdgpu_architectures)
+    set(gpu_target_triple ${AMDGPU_TARGET_TRIPLE})
+    list(APPEND compile_options "-mcpu=${gpu_arch}")
+    list(APPEND compile_options "SHELL:-Xclang -mcode-object-version=none")
```
Comment (SchrodingerZhu) [inline libc/cmake/modules/LLVMLibCObjectRules.cmake:177]: Out of curiosity, what does `SHELL:` prefix do here

Comment (jhuber6) [reply libc/cmake/modules/LLVMLibCObjectRules.cmake:177]: Normally CMake deduplicates command line arguments, that's bad for things like `-Xclang` or `-mllvm` that need to precede another argument. `SHELL:` make it not do that.

**Review by jhuber6 — COMMENTED** [review]
```
[736/992] Linking CXX executable libc/test/integration/startup/gpu/libc.test.integration.startup.gpu.init_fini_array_test.__build__
FAILED: libc/test/integration/startup/gpu/libc.test.integration.startup.gpu.init_fini_array_test.__build__ 
: && /home/jhuber/Documents/llvm/llvm-project/build/./bin/clang++ --target=x86_64-unknown-linux-gnu -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -fuse-ld=lld -Wl,--color-diagnostics    -nostdlib -static libc/src/__support/RPC/CMakeFiles/libc.src.__support.RPC.rpc_client.__internal__.dir/rpc_client.cpp.o libc/src/__support/OSUtil/gpu/CMakeFiles/libc.src.__support.OSUtil.gpu.gpu_util.__internal__.dir/quick_exit.cpp.o libc/src/__support/OSUtil/gpu/CMakeFiles/libc.src.__support.OSUtil.gpu.gpu_util.__internal__.dir/io.cpp.o libc/src/stdlib/CMakeFiles/libc.src.stdlib._Exit.__internal__.dir/_Exit.cpp.o libc/src/__support/CPP/CMakeFiles/libc.src.__support.CPP.new.__internal__.dir/new.cpp.o libc/src/stdlib/CMakeFiles/libc.src.stdlib.atexit.__internal__.dir/atexit.cpp.o libc/src/stdlib/CMakeFiles/libc.src.stdlib.exit.__internal__.dir/exit.cpp.o libc/startup/gpu/nvptx/CMakeFiles/libc.startup.gpu.nvptx.crt1.dir/start.cpp.o libc/test/IntegrationTest/CMakeFiles/libc.test.IntegrationTest.test.dir/test.cpp.o libc/src/string/CMakeFiles/libc.src.string.bcmp.__internal__.dir/bcmp.cpp.o libc/src/string/CMakeFiles/libc.src.string.bzero.__internal__.dir/bzero.cpp.o libc/src/string/CMakeFiles/libc.src.string.memcmp.__internal__.dir/memcmp.cpp.o libc/src/string/CMakeFiles/libc.src.string.memcpy.__internal__.dir/memcpy.cpp.o libc/src/string/CMakeFiles/libc.src.string.memmove.__internal__.dir/memmove.cpp.o libc/src/string/CMakeFiles/libc.src.string.memset.__internal__.dir/memset.cpp.o libc/test/integration/startup/gpu/CMakeFiles/libc.test.integration.startup.gpu.init_fini_array_test.__build__.dir/init_fini_array_test.cpp.o -o libc/test/integration/startup/gpu/libc.test.integration.startup.gpu.init_fini_array_test.__build__  -march=sm_89  --target=nvptx64-nvidia-cuda  --cuda-path=/opt/cuda && :
nvlink fatal   : cubin not an elf?
```
Getting these typical errors. Means something either isn't built with `--target=` or we picked up an incorrect file.

File: libc/cmake/modules/LLVMLibCObjectRules.cmake [hunk 2]
```diff
@@ -567,16 +588,25 @@ function(create_entrypoint_object fq_target_name)
...
       ${internal_target_name}
+      ${LIBC_GPU_TARGET_ARCHITECTURE}
       SRCS ${ADD_ENTRYPOINT_OBJ_SRCS}
       HDRS ${ADD_ENTRYPOINT_OBJ_HDRS}
       COMPILE_OPTIONS ${common_compile_options}
       CXX_STANDARD ${ADD_ENTRYPOINT_OBJ_CXX_STANDARD}
       DEPENDS ${full_deps_list}
       FLAGS "${ADD_ENTRYPOINT_OBJ_FLAGS}"
-    )
+    )
```
Comment (jhuber6) [inline libc/cmake/modules/LLVMLibCObjectRules.cmake:610]: Whitespace?

**Review by jhuber6 — APPROVED** [review]
Passes on both NVPTX and AMDGPU now. Thanks.

**Pull request merged**